### PR TITLE
feature(#166): add spreadProp & spreadPath

### PR DESCRIPTION
**Pull request by BjornMelgaard**

### CODE REVIEW

**Comment by char0n** [comment]
Will check during today

File: src/spreadPath.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Merges object under path with root object, removing path to object.
+ *
+ * @func spreadPath
+ * @memberOf RA
+ * @since {@link https://char0n.github.io/ramda-adjunct/1.19.0|v1.19.0}
+ * @category Object
+ * @sig [k] -> {a: *} -> {b: *}
+ * @param {!Object} path The path to object to merge with
+ * @param {!Object} obj The root object
```
Comment (char0n) [inline src/spreadPath.js]: Let's borrow the lingo from [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator)

`The provided object`

File: src/spreadPath.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { curry, pathOr, merge, dissoc } from 'ramda';
+
+/**
+ * Merges object under path with root object, removing path to object.
```
Comment (char0n) [inline src/spreadPath.js]: `Spreads object under property path onto provided object.`

IMHO we don't  need to tell that it removed the object from the path. Spread is by definition a new composition of the object.

File: src/spreadPath.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * @func spreadPath
+ * @memberOf RA
+ * @since {@link https://char0n.github.io/ramda-adjunct/1.19.0|v1.19.0}
+ * @category Object
+ * @sig [k] -> {a: *} -> {b: *}
```
Comment (char0n) [inline src/spreadPath.js]: Isn't this signature closer to actually what happens ?
```javascript
 * @sig
 *   [Idx] -> {k: v} -> {k: v}
 *   Idx = String | Int
```

File: test/spreadPath.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import RA from '../src/index';
+import eq from './shared/eq';
+
+
+describe('spreadPath', function() {
```
Comment (char0n) [inline test/spreadPath.js]: We should finally utilize the power of mocha more. I'll demonstrate what I have in mind.
Also don't try to create variables when they are not needed.

```javascript
import RA from '../src/index';
import eq from './shared/eq';


describe('spreadPath', function() {
  let obj;

  beforeEach(function () {
    obj =  {
      a: 1,
      b1: { b2: { c: 3, d: 4 } },
    };
  });

  it('tests currying', function () {
    const path = ['b1', 'b2'];
    const expected = { a: 1, c: 3, d: 4  };

    eq(RA.spreadPath(path, obj), expected);
    eq(RA.spreadPath(path)(obj), expected);
  });

  context('when path leads  to non object', function () {
    specify('should return object without path', function () {
      const obj = {
        a: 1,
        b1: { b2: 999 },
      };
      eq(RA.spreadPath(['b1', 'b2'], obj), { a: 1, b1: {} });    
    });
  });
  
  context("when path doesn't exist", function () {
    specify('should return object with identical structure as provided object', function () {
      eq(RA.spreadPath(['does', 'not', 'exist'], obj), obj);
    });
  });  
});

File: test/spreadProp.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import RA from '../src/index';
+import eq from './shared/eq';
+
+
+describe('spreadProp', function() {
```
Comment (char0n) [inline test/spreadProp.js]: Redo these tests according what I explained on `spreadPath` tests and I think we should have more tests here

 - [x] currying
 - [ ] non existing prop
 - [x] prop that is non object
 - [ ] test for overriding existing props on provided object
 - [ ] test for standard usecase

File: src/spreadProp.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import { of, curry } from 'ramda';
+
+import spreadPath from './spreadPath';
+
+/**
+ * Merges object under prop with root object, removing prop to object.
```
Comment (char0n) [inline src/spreadProp.js]: Spreads object under property onto provided object.

File: src/spreadProp.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * @func spreadProp
+ * @memberOf RA
+ * @since {@link https://char0n.github.io/ramda-adjunct/1.19.0|v1.19.0}
+ * @category Object
+ * @sig @sig k -> {a: *} -> {b: *}
```
Comment (char0n) [inline src/spreadProp.js]:
```javascript
 * @sig
 *   Idx -> {k: v} -> {k: v}
 *   Idx = String | Int
```

File: src/spreadProp.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ * @param {!Object} prop The name of property that holds object to merge with
```
Comment (char0n) [inline src/spreadProp.js]: The property path to spread

File: src/spreadProp.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ * @param {!Object} obj The root object
```
Comment (char0n) [inline src/spreadProp.js]: The provided object

File: src/spreadProp.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+ * @return {!Object} Result of merging root object with object under prop
```
Comment (char0n) [inline src/spreadProp.js]: The result of the spread

File: src/spreadProp.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+ * @see {@link RA.spreadPath|spreadPath}
+ * @example
+ *
+ * R.spreadProp("b", { a: 1, b: { c: 3, d: 4 } }); // => { a: 1, c: 3, d: 4 }
```
Comment (char0n) [inline src/spreadProp.js]: use single quotes: 'b'

File: src/spreadProp.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
```
Comment (char0n) [inline src/spreadProp.js]: drop this empty line

File: test/spreadProp.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  it('tests currying', function () {
+    const prop = 'b';
```
Comment (char0n) [inline test/spreadProp.js]: line 7 to 14 are not needed here IMHO. We are testing currying here only. The body of this test could be following
```javascript
it('tests curring', function () {
  eq(RA.spreadProp('prop', {}), {}); 
  eq(RA.spreadProp('prop')({}), {});
});

File: src/spreadPath.js [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+ * Spreads object under property path onto provided object.
+ *
+ * @func spreadPath
+ * @memberOf RA
+ * @since {@link https://char0n.github.io/ramda-adjunct/1.19.0|v1.19.0}
+ * @category Object
+ * @sig
+ *   [Idx] -> {k: v} -> {k: v}
+ *   Idx = String | Int
+ * @param {!Object} path The property path to spread
```
Comment (char0n) [inline src/spreadPath.js]: ```@param {!Array.<string|number>}``` Not sure about proper syntax for union types in jsdoc

File: src/spreadPath.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+ * @param {!Object} obj The provided object
+ * @return {!Object} The result of the spread
+ * @see {@link RA.spreadProp|spreadProp}
+ * @example
+ *
+ * R.spreadPath(
+ *   ['b1', 'b2'],
+ *   { a: 1, b1: { b2: { c: 3, d: 4 } } }
+ * ); // => { a: 1, c: 3, d: 4, b1 {} }
```
Comment (char0n) [inline src/spreadPath.js]: missing colon `{ a: 1, c: 3, d: 4, b1: {} }`

File: src/spreadProp.js [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+ * Spreads object under property onto provided object.
+ *
+ * @func spreadProp
+ * @memberOf RA
+ * @since {@link https://char0n.github.io/ramda-adjunct/1.19.0|v1.19.0}
+ * @category Object
+ * @sig
+ *   Idx -> {k: v} -> {k: v}
+ *   Idx = String | Int
+ * @param {!Object} prop The property to spread
```
Comment (char0n) [inline src/spreadProp.js]: ```@param {string|number}```

**Comment by BjornMelgaard** [comment]
done

**Pull request merged**